servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/SingleThreadedExecutor.java

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright © 2021 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.concurrent.api;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.ref.WeakReference;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.SynchronousQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+import java.util.function.Supplier;
+
+import static java.util.concurrent.atomic.AtomicIntegerFieldUpdater.newUpdater;
+
+/**
+ * Executor that can execute one task at a time on a dedicated worker thread. While concurrent tasks are generally
+ * rejected, the `execute(Runnable)` method provides special behavior if the invoking thread is the executor worker
+ * thread: the `Runnable` will be executed immediately on the calling thread.
+ */
+class SingleThreadedExecutor implements java.util.concurrent.Executor, AutoCloseable {
```

Comment:
Not sure about `java.io.Closeable` as it has nothing to do with IO.
Consider overriding `close()` method with more appropriate exception based on what we learned from a use-case with grpc #1566

Reply:
Since it is not a public API class I don't find the need to mark it (or the thread subclass) as `final`. The historical performance reasons will be handled by the JIT so dispatch should be no slower that if it were declared `final`.